PHOENIX-6611 Multiple IndexTool bugs

stoty opened the pull request:
Copying from the ticket:

There is a lot of noise, due to the removal of the directApi option, and movinf the VERIFIED/UNVERIFIED constants to QueryConstants.
The interesting stuff is in:

IndexTool
FormatToBytesWritableMapper
PhoenixIndexImportDirectMapper
PhoenixTxIndexMutationGenerator
IndexUtil
IndexMaintainer
IndexToolIT
IndexBuildTimestampIT

stoty commented:
Sorry for asking for reviews for a half-done patch.

It is really ready for review now. 
(The last patch is just just formatting fixes, and does not change the behaviour.)

gjacoby126 commented on phoenix-core/src/main/java/org/apache/phoenix/mapreduce/index/IndexTool.java:
```diff
@@ -493,6 +493,10 @@ public Job getJob() throws Exception {
                 if (useSnapshot || (!isLocalIndexBuild && pDataTable.isTransactional())) {
                     PhoenixConfigurationUtil.setCurrentScnValue(configuration, maxTimeRange);
+                    if (indexVerifyType != IndexVerifyType.NONE) {
+                        LOGGER.warn("Verification is not supported for client index rebuilds,"
```
The error message seems wrong -- verification is supported for some client-initiated index rebuilds, just not ones using a mapper other than the server build one. If you mean that the direct mapper creates the mutations and writes them as a client, that's probably not clear to the end user.

gjacoby126 reviewed (changes requested):
@stoty - Very elegant solution to fixing the direct mapper. Just a small nit.

gokceni commented on phoenix-core/src/main/java/org/apache/phoenix/execute/PhoenixTxIndexMutationGenerator.java:
```diff
@@ -364,7 +364,8 @@ private boolean generatePuts(
         }
-        Iterable<IndexUpdate> puts = codec.getIndexUpserts(state, indexMetaData, regionStartKey, regionEndKey);
+        Iterable<IndexUpdate> puts =
+                codec.getIndexUpserts(state, indexMetaData, regionStartKey, regionEndKey, true);
```
Shouldn't verified be false here? Are we generating index upsert mutations in verified=true even though the data table write fails? A code comment would be nice.

gjacoby126 replied:
@gokceni - I believe the purpose is to allow the old "direct" index tool mapper to be able to build new-style indexes, since snapshot-based and transactional index builds still need to use the direct mapper. Since it's reading existing base table data, we can assume the new index write is verified.

stoty replied:
This is in the normal write path via PhoenixTransactionalIndexer, not the in the async index generation.

Three-phase commit is not used in the normal TX write path, because the transactionality is taking care of the
consistency, and accordingly it never updates the status to VERIFIED, and leaves the status as "x".

This method is also used by the old indexing implementation, and I didn't want old-style tables to set the status to verified, hence the new parameter.

The same basic problem pattern that BulkLoad and Async Indextool has, but this one is in the normal write path.

I should have called this change out explicitly, as admittedly this has the bigggest impact in the patch, but forgot, sorry.

This really should have been four patches (snapshot, Async VERIFIED, TX index status, and the directApi purge), but I got carried away as I kept finding the issues by the expanded test case.

gokceni commented on phoenix-core/src/it/java/org/apache/phoenix/end2end/IndexToolIT.java:
```diff
@@ -378,6 +386,20 @@ public void testSecondaryIndex() throws Exception {
         }
     }
 
+    public static void checkVerifiedStatus(Connection conn, String indexFullName, byte[] rowKey) throws SQLException, IOException {
```
ImmutableIndexIT has a similar method (verifyRowsForEmptyColValue). Could we reuse it here?

stoty replied:
I've merged all three similar methods that I could find.

gokceni reviewed (commented):
Left some comments..

stoty commented:
Thank you @gjacoby126  and @gokceni.

I have implemented your suggestions, and explained the TX index change.

gokceni commented:
@stoty Thank you for implementing suggestions. There seems to be unit test failures. Could you please check if they are related to this patch? It LGTM but please make sure that tests are not broken.

gjacoby126 reviewed (approved):
+1, thanks @stoty (I'm assuming the permissions failures are due to them being flaky, but would be good to verify they pass locally before merging.)

stoty commented:
@gokceni 
I had to rebase this on top your TransformTool work.
For now, I have added logic in TransformMaintainer to set the VERIFIED status according to the old table indexing style.

Please check if that is acceptable solution (at least until PHOENIX-6617)

joshelser commented on phoenix-core/src/main/java/org/apache/phoenix/mapreduce/index/IndexTool.java:
```diff
@@ -667,7 +672,12 @@ private Job configureJobForAsyncIndex() throws Exception {
+                            .append("-Snapshot-")
+                            .append(System.currentTimeMillis())
+                            .toString();
+                    //FIXME Drop this snapshot after we're done ?
```
👍🏼

stoty replied:
I think I'm going to open a follow-up ticket for that, this patch already has more than enough changes.

joshelser commented on phoenix-core/src/main/java/org/apache/phoenix/schema/transform/TransformMaintainer.java:
```diff
@@ -133,6 +140,30 @@ private TransformMaintainer(final PTable oldTable, final PTable newTable, Phoeni
+                    newTableIndexVerified = true;
+                }
+            }
+        } catch (IOException e) {
+            throw new IllegalArgumentException("Could not determine empty column value", e);
+        } catch (SQLException e) {
```
nit: mulit-catch `catch (IOException | SQLException e) {...`

stoty replied:
Thanks, done.

joshelser reviewed (approved):
I'm a bit out of my depth (obviously, given the review feedback I can give) but looks fine to me :)

stoty commented:
I have removed the VERIFIED I added to TransformMaintainer, as it is not necessary after PHOENIX-6617.

I think that it would be a good idea to check if Transform handles the VERIFED status for transactional tables correctly, to avoid issues like IndexTool and Bulk Load had.
